p2panda-rs/src/schema/field_types.rs

```diff
@@ -19,115 +32,134 @@ pub enum FieldType {
-    PinnedRelationList,
+    /// Defines a [`PinnedRelationList`][`crate::operation::PinnedRelationList`] field that
+    /// references the given schema.
+    PinnedRelationList(SchemaId),
 }
 
 impl FieldType {
-    /// Returns the string representation of this type.
-    pub fn as_str(&self) -> &str {
+    /// Serialises this field type to text.
+    pub fn serialise(&self) -> String {
         match self {
-            FieldType::Bool => "bool",
-            FieldType::Int => "int",
-            FieldType::Float => "float",
-            FieldType::String => "str",
-            FieldType::Relation => "relation",
-            FieldType::RelationList => "relation_list",
-            FieldType::PinnedRelation => "pinned_relation",
-            FieldType::PinnedRelationList => "pinned_relation_list",
+            FieldType::Bool => "bool".to_string(),
+            FieldType::Int => "int".to_string(),
+            FieldType::Float => "float".to_string(),
+            FieldType::String => "str".to_string(),
+            FieldType::Relation(schema_id) => format!("relation({})", schema_id.as_str()),
+            FieldType::RelationList(schema_id) => format!("relation_list({})", schema_id.as_str()),
+            FieldType::PinnedRelation(schema_id) => {
+                format!("pinned_relation({})", schema_id.as_str())
+            }
+            FieldType::PinnedRelationList(schema_id) => {
+                format!("pinned_relation_list({})", schema_id.as_str())
+            }
         }
     }
 }
 
+impl From<FieldType> for OperationValue {
+    fn from(field_type: FieldType) -> OperationValue {
+        OperationValue::Text(field_type.serialise())
+    }
+}
+
 impl FromStr for FieldType {
     type Err = FieldTypeError;
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        match s {
-            "bool" => Ok(FieldType::Bool),
-            "int" => Ok(FieldType::Int),
-            "float" => Ok(FieldType::Float),
-            "str" => Ok(FieldType::String),
-            "relation" => Ok(FieldType::Relation),
-            "relation_list" => Ok(FieldType::RelationList),
-            "pinned_relation" => Ok(FieldType::PinnedRelation),
-            "pinned_relation_list" => Ok(FieldType::PinnedRelationList),
-            type_str => Err(FieldTypeError::InvalidFieldType(type_str.into())),
+        // Matches a field type name, followed by an optional group in parentheses that contains the
+        // referenced schema for relation field types.
+        let re = Regex::new(r"(\w+)(\((.+)\))?").unwrap();
+        let groups = re.captures(s).unwrap();
+
+        match (
+            groups.get(1).map(|m| m.as_str()),
+            groups.get(3).map(|m| m.as_str()),
+        ) {
+            (Some("bool"), None) => Ok(FieldType::Bool),
+            (Some("int"), None) => Ok(FieldType::Int),
+            (Some("float"), None) => Ok(FieldType::Float),
+            (Some("str"), None) => Ok(FieldType::String),
```

Comment:
It would be nice to only use regex when it is needed (for relations), for most cases we could simply just match by string.

Reply:
I don't care either way, I thought it was simpler here to just do everything by regex. Feel free to make changes if you want though, because i don't have time for that this week :)

Reply:
Done